### ISSUE TITLE
Checkboxes aren't converted properly

### ISSUE THREAD

**Issue by sojusnik** [peter88213/zim2obsidian#13]
**Application/script**
- Version: v0.6.6

**Describe the bug**
At the moment checkboxes are converted to `* ☐` after Zim's markdown export (this script seems not to alter them), but they actually should be converted to `- [ ]` to be rendered properly and be clickable in Obsidian. And `* ☑` checked off checkboxes should be formatted to `- [x]`.

Zim has 3 states of checkboxes that Obsidian isn't supporting:
```
[x] cross (note that this checkbox marks dismissed, or whatever you call them, checkboxes, while checked off checkboxes are written as [*] in Zim, so don't confuse it with checked off checkboxes in Obsidian, which are written as `- [x]`
[>] arrow right
[<] arrow left
```

Since Obsidian will still render checkboxes with a letter or sign between the square brackets (like `- [c]` or `- [>]`) as an actionable checkbox, but in this case only the checkbox itself will be filled and the appropriate text won't be crossed out, as it would be the case with a `- [x]` checkbox, I still suggest converting those checkboxes as follows:

`* ☒` cross should be converted to `- [c]` (because x is already taken and X might lead to confusion)
`* ▷` arrow right should be converted to `- [>]`
`* ◁` arrow left should be converted to `- [<]`

**Execution environment (please complete the following information):**
- Python Version: Python 3.11.2
- OS: Ubuntu 23.04

**Comment by peter88213** [comment]
This is how Zim's markdown export looks like:
```
* ☐ Unchecked
* ☑ Checked
* ☒ Cross-checked
* ▷ Arrow right
* ◁ Arrow left
```

To replace with: 

```
- [ ] Unchecked
- [*] Checked
- [c] Cross-checked
- [>] Arrow right
- [<] Arrow left
```

The replacements can be done in various ways. 
My suggestion: 

```
checkboxes = {
                 '* ☐' : '- [ ]',
                 '* ☑' : '- [*]',
                 '* ☒' : '- [c]',
                 '* ▷' : '- [>]',
                 '* ◁' : '- [<]',
                  }

for c in checkboxes:
    if line.startswith(c):
        line = f'{checkboxes[c]}{line[3:]}'
```

**Comment by peter88213** [comment]
Right. I fixed the comment, sorry. However, the code is correct as you can see. 

I would really like to encourage you to learn Python. As you can see, It's a language made for text processing. 
The changes for v0.8.0 are a 10-minutes-job, and the code is a beginner's exercise. 

I implemented the change using the test driven development method: 
First I prepared the [test files](https://github.com/peter88213/zim2obsidian/tree/main/test/data), so the specification was set. Then I implemented the algorithm, and running the [test script](https://github.com/peter88213/zim2obsidian/blob/main/src/test_zim2obsidian.py), I made sure that it works.

**Comment by sojusnik** [comment]
> I would really like to encourage you to learn Python. As you can see, It's a language made for text processing.

I already have too much going on right now, so learning Python, even on a basic level, would worsen the situation even more. Although I could imagine learning it in the near future, if time will permit.

There are still issues with those checkboxes:

1. If checkboxes are indented, then conversion fails, see:

```
[ ] checkboxes aren't converted properly when indented
	[ ] example
		[ ] another example
```
In this example, the second and third checkbox won't be converted properly. Might be connected to https://github.com/peter88213/zim2obsidian/issues/11

2. Zim's markdown export adds a new line before and after each line with a checkbox, except between consecutive checkboxes, see the following:

```
==== Header 3 ====
[ ] before this line a new line will be created
[ ] example
[ ] after this line a new line will be created

Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
[ ] before and after this line a new line will be created

Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
	[ ] and before this indented checkbox, a new line will be created before and after this line 
```

**Comment by peter88213** [comment]
In order not to get confused, we have to strictly distinguish between what Zim can (and cannot) export and the function of the *zim2obsidian* postprocessor. 

As you can see from the code, the postprocessor only converts checkboxes at the beginning of lines that are preceded by a bullet. This is the way it is with Zim's to-do lists. At the moment I just don't see any sense in having several checkboxes on one line, or indented checkboxes, but who wants to can of course extend the post-processor's behavior at will. 

I haven't yet tried whether Obsidian lets you indent lines with tabs at all; Markdown, as I mentioned, doesn't provide for that. It's also not my style to write that way. 

If Zim allows more formatting options than standard Markdown, it's just not fair to call it a bug of the conversion if everything doesn't transfer one-to-one to Obsidian.

**Comment by sojusnik** [comment]
> I haven't yet tried whether Obsidian lets you indent lines with tabs at all; Markdown, as I mentioned, doesn't provide for that. It's also not my style to write that way.

It does:
![grafik](https://github.com/peter88213/zim2obsidian/assets/1245718/75e8b7a7-fd9b-492a-b36e-9d5147f2e4a4)

Written in Markdown in Obsidian:
```
- [ ] example
	- [x] example
		- [ ] example
```

> As you can see from the code, the postprocessor only converts checkboxes at the beginning of lines that are preceded by a bullet. This is the way it is with Zim's to-do lists. At the moment I just don't see any sense in having several checkboxes on one line, or indented checkboxes, but who wants to can of course extend the post-processor's behavior at will.

So I assume the proper way would be to report those issues in Zim?

**Comment by peter88213** [comment]
> So I assume the proper way would be to report those issues in Zim?

In Zim, feature requestsd should only be reported under "Issues" if you have finished code to offer for solution. 

If you want formatting converted that works in both Zim and Obsidian, but doesn't transfer via standard Markdown, the correct approach would be @pschwede's. Perhaps someone will be found to further develop his script. The difficulty I see with this is that he has not included any copyright or licensing information. One could consider his code as "public domain", but there is no such thing under German law (I guess, he's German).

What you can definitely do is post your requirements for his script as Issues. But be polite: call it a bug only if the program crashes or makes complete nonsense. If something you want is not converted, call it "missing feature".

**Comment by sojusnik** [comment]
> But be polite: call it a bug only if the program crashes or makes complete nonsense. If something you want is not converted, call it "missing feature".

Wasn't aware that writing feedback in a bug report could be perceived as impolite, if you take the current issue as an example. That was not my intention. I was looking at it more from a point of view of efficiency, having everything related in one place.

> If you want formatting converted that works in both Zim and Obsidian, but doesn't transfer via standard Markdown, the correct approach would be @pschwede's.

Your awesome script already does nearly a perfect job. Only the issues with [indentations](https://github.com/peter88213/zim2obsidian/issues/11) and the checkboxes are left and we are good to go. So in this situation, switching to another approach would be counterintuitive. And the elegance of your approach is already unmatched.

**Comment by peter88213** [comment]
Well, as we know now, Indentiations with tab characters cannot transferred via Markdown.
Tabs are not part of the Markdown specification. So there's no point in asking for tab support in Zim's Markdown export.

What I could imagine would be a "dirty hack". One could edit the original Zim notebook with a preprocessor and replace all tab characters with a special code. Then export with *Zim* and post-process with *zim2obsidian*. Then at the very end, have the special codes in the Markdown files replaced with tab characters again. 
The necessary scripts could easily be generated by modifying *zim2obsidian* functions.

**Comment by sojusnik** [comment]
In continuation of [this](https://github.com/peter88213/zim2obsidian/issues/13#issuecomment-1648373657) issue regarding the appearance of new empty lines:

Currently, Zim's markdown export adds new empty lines before and after already present text lines starting with checkboxes, numbered and unnumbered lists. Take this Zim source as example describing where new lines will appear:

```
==== Header 3 ====
[ ] before this line a new line will be created
[ ] example
[ ] after this line a new line will be created

Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
[ ] before and after this line a new line will be created

Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
	[ ] and before this indented checkbox, a new line will be created before and after this line

Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
1. before and after this line a new line will be created

Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
* before and after this line a new line will be created
```

Do you think that this should be reported upstream in Zim or maybe there could be also a hackish solution to it, like done with indentations?

**Comment by peter88213** [comment]
I think this is because in Markdown a line break is represented by a blank line. 
Bullet points are the exception, but checkboxes don't count as bullet points in Markdown. 
If you indent checkboxes in Zim so that there is no bullet at the beginning of the line, the Markdown exporter will correctly insert a blank line. 
Whatever way you want to solve this, it is not an issue for Zim.

**Comment by peter88213** [comment]
I guess, Zim's Markdown exporter inserts the bullet points only before checkboxes at the beginning of a line, in order to generate a Markdown To-Do-List. 

Processing multi-level check lists turns out to be a complicated task. At the moment, there is probably nothing left to do but rework manually.

**Comment by sojusnik** [comment]
Bad news, but let's see what time will bring. Maybe somebody else has an idea.

Thanks for the work you've already done!
